flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java

```diff
@@ -0,0 +1,207 @@
+        // size threshold could be added to call persist when reached. considerations:
+        // 0. can actually degrade performance by amplifying number of requests
+        // 1. which range to persist?
+        // 2. how to deal with retries/aborts?
+    }
+
+    @Override
+    public SequenceNumber lastAppendedSequenceNumber() {
+        rollover();
+        LOG.trace("query {} sqn: {}", logId, lastAppendedSequenceNumber);
+        return lastAppendedSequenceNumber;
+    }
+
+    @Override
+    public CompletableFuture<StateChangelogHandleStreamImpl> persist(SequenceNumber from)
+            throws IOException {
+        LOG.debug("persist {} from {}", logId, from);
+        checkNotNull(from);
+        // todo: check range
+
+        rollover();
+        Collection<StoreResult> readyToReturn = confirmed.tailMap(from, true).values();
+        Collection<StateChangeSet> toUpload = changeSets.tailMap(from, true).values();
+        LOG.debug("collected readyToReturn: {}, toUpload: {}", readyToReturn, toUpload);
+        CompletableFuture<StateChangelogHandleStreamImpl> future = new CompletableFuture<>();
+        store.save(createTask(readyToReturn, toUpload, future));
+        return future;
+    }
+
+    private StoreTask createTask(
+            Collection<StoreResult> readyToReturn,
+            Collection<StateChangeSet> toUpload,
+            CompletableFuture<StateChangelogHandleStreamImpl> future) {
+        return new StoreTask(
+                toUpload,
+                future::completeExceptionally,
+                results ->
+                        executor.execute(
+                                () -> {
+                                    results.forEach(e -> uploaded.put(e.sequenceNumber, e));
+                                    future.complete(buildHandle(results, readyToReturn));
+                                }));
+    }
+
+    @Override
+    public void close() {
+        LOG.debug("close {}", logId);
+        checkState(!closed);
+        closed = true;
+        activeChangeSet.clear();
+        // todo in MVP or later: cleanup if transition succeeded and had non-shared state
+        changeSets.clear();
+        uploaded.clear();
+        confirmed.clear();
+        // the store is closed from the owning FsStateChangelogClient
+    }
+
+    @Override
+    public void confirm(SequenceNumber from, SequenceNumber to) {
+        LOG.debug("confirm {} from {} to {}", logId, from, to);
+        uploaded.subMap(from, true, to, false)
+                .forEach(
+                        (sequenceNumber, changeSetAndResult) -> {
+                            changeSets.remove(sequenceNumber);
+                            uploaded.remove(sequenceNumber);
+                            confirmed.put(sequenceNumber, changeSetAndResult);
+                        });
+    }
+
+    @Override
+    public void reset(SequenceNumber from, SequenceNumber to) {
+        LOG.debug("reset {} from {} to {}", logId, from, to);
```

Comment:
What if we maintain the status of the changelogset ?  we can just change the status of them if we confirm or reset .

Reply:
I guess that's not relevant after adopting TM-side state ownership (`reset` is now no-op).